Source/CarthageKit/FrameworkExtensions.swift

```diff
@@ -154,3 +155,72 @@ extension NSScanner {
 		return nsString.substringWithRange(lineRange)
 	}
 }
+
+extension NSURLSession {
+	/// Returns a signal that will download a file using the given request. The
+	/// file will be deleted after the URL has been sent upon the signal.
+	internal func rac_downloadWithRequest(request: NSURLRequest) -> ColdSignal<(NSURL, NSURLResponse)> {
+		return ColdSignal { (sink, disposable) in
+			let serialDisposable = SerialDisposable()
+			let handle = disposable.addDisposable(serialDisposable)
+
+			let task = self.downloadTaskWithRequest(request) { (URL, response, error) in
+				// Avoid invoking cancel(), or the download may be deleted.
+				handle.remove()
+
+				if URL == nil || response == nil {
+					sink.put(.Error(error))
+				} else {
+					let value = (URL!, response!)
+					sink.put(.Next(Box(value)))
+					sink.put(.Completed)
+				}
+			}
+
+			serialDisposable.innerDisposable = ActionDisposable {
+				task.cancel()
+			}
+
+			task.resume()
+		}
+	}
+}
+
+extension NSURL: JSONDecodable {
+	public class func decode(json: JSONValue) -> Self? {
+		if let URLString = String.decode(json) {
+			return self(string: URLString)
+		} else {
+			return nil
+		}
```

Comment:
> This is the same method we talked about before. map won't work because it would need the signature of `>>=`.

Postfix `?` turns out to be monadic `join`, so:

```swift
return String.decode(json).map { self(string: $0) }?
```

should work. But I’m not bothered if you want to leave it as is.

> This is specifically adding conformance to an Argo protocol, so it has to match their signature.

Ah, of course.